cmd/minio-cosi-driver/internal/provisioner.go

```diff
@@ -0,0 +1,119 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/container-object-storage-interface-provisioner-sidecar/cmd/minio-cosi-driver/internal/minio"
+	cosi "sigs.k8s.io/container-object-storage-interface-spec"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type ProvisionerServer struct {
+	provisioner string
+	mc          *minio.C
+}
+
+// ProvisionerCreateBucket is an idempotent method for creating buckets
+// It is expected to create the same bucket given a bucketName and protocol
+// If the bucket already exists, then it MUST return codes.AlreadyExists
+// Return values
+//    nil -                   Bucket successfully created
+//    codes.AlreadyExists -   Bucket already exists. No more retries
+//    non-nil err -           Internal error                                [requeue'd with exponential backoff]
+func (s *ProvisionerServer) ProvisionerCreateBucket(ctx context.Context,
+	req *cosi.ProvisionerCreateBucketRequest) (*cosi.ProvisionerCreateBucketResponse, error) {
+
+	bucketName := req.GetName()
+	klog.V(3).InfoS("Create Bucket", "name", bucketName)
+
+	protocol := req.GetProtocol()
+	if protocol == nil {
+		klog.ErrorS(errors.New("Invalid Argument"), "Protocol is nil")
+		return nil, status.Error(codes.InvalidArgument, "Protocol is nil")
+	}
+	s3 := protocol.GetS3()
+	if s3 == nil {
+		klog.ErrorS(errors.New("Invalid Argument"), "S3 protocol is nil")
+		return nil, status.Error(codes.InvalidArgument, "S3 Protocol is nil")
+	}
+
+	options := minio.MakeBucketOptions{}
+
+	// MinIO regions, unlike AWS s3 does not strictly require the
+	// country-direction-index format. Therefore, no validation
+	// is needed here
+	options.Region = s3.Region
+
+	// Support for the following two fields will be added
+	// in the future using which bucket will be provisioned in a
+	// particular region conforming to a particular signature version
+	// However, as of now, these will be ignored
+
+	// endpoint := s3.Endpoint
+	// signatureVersion := s3.SignatureVersion
+
+	// Since 'parameters' is not a typed construct
+	// it is better to have predefined set of keys
+	// to parse, rather than treating it as an opaque
+	// set of keys and values.
+	parameters := req.GetParameters()
+
+	for k, v := range parameters {
+		switch k {
+		case minio.ObjectLocking:
+			options.ObjectLocking = true
+		default:
+			klog.ErrorS(errors.New("Invalid Argument"), "parameter", k, "value", v)
+			return nil, status.Error(codes.InvalidArgument, "invalid parameter")
+		}
+	}
+
+	bucketID, err := s.mc.CreateBucket(ctx, bucketName, options)
+	if err != nil {
+		if err == minio.ErrBucketAlreadyExists {
+			klog.InfoS("Bucket already exists", "name", bucketName)
+			return nil, status.Error(codes.AlreadyExists, "Bucket already exists")
```

Comment:
If the bucket already exists, then will return nil which would cause a nil pointer exception here: https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar/blob/master/pkg/bucket/bucket_controller.go#L108